### PR TITLE
Implementation of deltaspin

### LINKED ISSUE #1851: Constrained magnetization for noncolinear spin
### Background

This is for collaboration with DeepH. We might consider implement non-collinear spin-constrained DFT which is available in both SIESTA and OpenMX.

### Describe the solution you'd like

Details of the method might be found in the following two references:

[Cuadrado_2018_J._Phys._Mater._1_015010.pdf](https://github.com/deepmodeling/abacus-develop/files/10637277/Cuadrado_2018_J._Phys._Mater._1_015010.pdf)

[Cai, Z., & Xu, B. (2022). First-principles Study of Non-collinear Spin Fluctuations Using Self-adaptive Spin-constrained Method. arXiv 2208.04551.pdf](https://github.com/deepmodeling/abacus-develop/files/10637278/Cai.Z.Xu.B.2022.First-principles.Study.of.Non-collinear.Spin.Fluctuations.Using.Self-adaptive.Spin-constrained.Method.arXiv.2208.04551.pdf)

### Additional Context

_No response_

### LINKED ISSUE #2488: How to constrain the magnetic orders in soc calculations
### Background

Estimating the magnetic ground state with the SOC effect is essential to investigate the magnetic properties in material science.

### Describe the solution you'd like

I want to compare several magnetic order with the consideration of SOC calculations, but I found ABACUS do not support this now.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by hongriTianqi**
### Reminder
- [x] Have you linked an issue with this pull request?
- [x] Have you noticed possible changes of behavior below or in the linked issue?
- [x] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
Fix #2488 
Fix #1851  

### What's changed?
- The function of deltaspin to constrain atomic magnetic moments in non-collinear calculation is added, which is based on the method published at [arXiv:2208.04551v6](https://arxiv.org/abs/2208.04551) by Zefeng Cai, Ben Xu, et al.

### Any changes of core modules? (ignore if not applicable)

All the changes below are active only when `GlobalV::sc_mag_switch` is true, which is controlled by the `INPUT` parameter of `sc_mag_switch`

#### ESolver_KS_LCAO
- `SpinConstrain::init_sc` is added in `ESolver_KS_LCAO::beforesolver` to initialize class SpinConstrain
- `SpinConstrain::run_lambda_loop` is called in `ESolver_KS_LCAO::eachiterinit` to run inner lambda loop
- `SpinConstrain::cal_MW` is called in `ESolver_KS_LCAO::hamilt2density` in order to calculate E_spin_constrain, which is needed in updating the Harris energy.
- `SpinConstrain::cal_MW` is called in `ESolver_KS_LCAO::heachiterfinish` to calculate atomic magnetic moment after each iteration finishes, which is needed in updating total energy.
- `SpinConstrain::cal_MW` is called in `ESolver_KS_LCAO::hafterscf` to calculate atomic magnetic moment after scf finishes

#### OperatorLCAO
- a new operator called `OperatorScLambda` is added in `OperatorLCAO` in order to calculate h_lambda

#### ElecstateLCAO
- `ElecStateLCAO<std::complex<double>>::get_spin_constrain_energy()` is added to calculate energy due to spin constraints

#### ElecState
- a new energy term `escon` in added in `struct fenergy` in to add the spin constraining energy in `ElecState: this->f_en.escon`
- the function of `calEBand()` is made public in order to be called in `SpinConstrain::cal_mw_from_lambda`.

### Unit test coverage results:
test1
test2

### CODE REVIEW

File: source/Makefile [hunk 1]
```diff
@@ -49,7 +49,7 @@ else
 endif
 
 ifeq ($(INTEL), ON)
-    OPTS += -lifcore -simd -qopenmp
+    OPTS += -lifcore  -qopenmp
```
Comment (dyzheng) [inline source/Makefile:52]: Why do you delete command of "-simd"?

File: source/module_io/input.h [hunk 2]
```diff
@@ -559,6 +559,21 @@ class Input
...
 
     //==========================================================
+    // variables for non-collinear spin-constrained DFT (deltaspin)
+    //==========================================================
+    /**
+     * 0: none spin-constrained DFT;
+     * 1: constrain atomic spin;
+     */
+    bool sc_mag_switch;
+    bool decay_grad_switch;
```
Comment (dyzheng) [inline source/module_io/input.h]: Annotation for this variable?

File: source/module_io/input_conv.h [hunk 3]
```diff
@@ -14,6 +14,7 @@
 #include <iostream>
 #include <string>
 #include <vector>
+#include <map>
```
Comment (dyzheng) [inline source/module_io/input_conv.h]: Why do you add this line?

File: source/module_cell/unitcell.cpp [hunk 4]
```diff
@@ -301,6 +301,47 @@ void UnitCell::set_iat2itia(void)
     return;
 }
 
+std::map<int, int> UnitCell::get_atomCounts() const
+{
+	std::map<int, int> atomCounts;
+	for (int it = 0; it < this->ntype; it++)
+	{
+		atomCounts.insert(std::pair<int, int>(it, this->atoms[it].na));
```
Comment (dyzheng) [inline source/module_cell/unitcell.cpp:309]: It seems a simple array can do this work, why do you use a map?

File: source/module_base/global_variable.h [hunk 5]
```diff
@@ -11,6 +11,7 @@
 #include <sstream>
 #include <string>
 #include <vector>
+#include "vector3.h"
```
Comment (dyzheng) [inline source/module_base/global_variable.h]: Why do you add this line?

File: source/module_elecstate/elecstate_lcao_tddft.h [hunk 6]
```diff
@@ -29,6 +29,7 @@ class ElecStateLCAO_TDDFT : public ElecStateLCAO<std::complex<double>>
     }
     void psiToRho_td(const psi::Psi<std::complex<double>>& psi);
     void calculate_weights_td();
+
```
Comment (dyzheng) [inline source/module_elecstate/elecstate_lcao_tddft.h]: Would you like to drop this change?

File: source/module_esolver/esolver_ks_lcao.cpp [hunk 7]
```diff
@@ -555,6 +557,12 @@ namespace ModuleESolver
         // update real space Hamiltonian
         this->p_hamilt->refresh();
     }
+
+    if (GlobalV::sc_mag_switch && iter > 1)
```
Comment (dyzheng) [inline source/module_esolver/esolver_ks_lcao.cpp:562]: Would you like to add some annotations in esolver for developers to read?

File: source/Makefile (hunk 1 above)
Comment (hongriTianqi) [reply source/Makefile:52]: I met compilation error with the option. @dyzheng

**Review by dyzheng — APPROVED** [review]
Congratulations!

File: source/module_esolver/esolver_ks_lcao.cpp (hunk 7 above)
Comment (hongriTianqi) [reply source/module_esolver/esolver_ks_lcao.cpp:562]: done

File: source/module_elecstate/elecstate_lcao_tddft.h (hunk 6 above)
Comment (hongriTianqi) [reply source/module_elecstate/elecstate_lcao_tddft.h]: done

File: source/module_base/global_variable.h (hunk 5 above)
Comment (hongriTianqi) [reply source/module_base/global_variable.h]: removed

**Pull request merged**